### ISSUE TITLE
Truncate or zero out log files

### ISSUE THREAD

**Issue by gbnewby** [gutenbergtools/ebookconverter#51]
Logfiles in cache/epub and ~gutenbackend/ebookconverter and perhaps elsewhere have grown quite large.

It should be safe to zero out the logfiles in cache/epub, and perhaps we should set a size limit of 1MB or less so old records are cut with every rebuild cycle.

For the ebookconverter logs, those could be mv'd and then new logfiles started. Should we set up a logrotate process to manage this? I'm not sure how long we should retain those logs for - perhaps 1 year, but we could do a monthly rotation.

**Comment by tangledhelix** [comment]
These logs are left behind by discrete runs, not long-running daemons, correct? I would suggest something different than file truncation. Each run can write its own discrete log with an epoch timestamp in the filename, and we could have a job come around and clean up any files with an mtime older than {something} periodically. Where {something} is whatever we like; 6 months, 1 month, 1 week... with this method, rather than truncation, you won't have files which start in the middle of a run because that's where some arbitrary line or byte count fell to start the file.

A secondary benefit of this is you'd be able to tell where the job began and ended. The current log smooshes them against one another and (for me anyway) makes it harder to read and determine what went on in a single invocation.

**Comment by eshellman** [comment]
That could work.

**Comment by tangledhelix** [comment]
> That could work.

How would we want to reap them? Keep _n_ generations? Keep them for _n_ days?

**Comment by cpeel** [comment]
If these are the logs I'm thinking of, they include a shitton of debug information as well. Is all of that necessary? Can we be _smarter_ about what we're logging?

**Comment by tangledhelix** [comment]
Maybe we could be smarter, but that doesn't solve the overall problem if we also don't rotate the log data and let it accumulate. It would just take longer to get to the same problem... I don't know how much space we'd save by doing per-run log files that are deleted after some period of time, but it's probably a lot, and it would deal with itself from then on. I think something like this paired with a reaper that gets rid of files after 100 days would do a lot to address the problem, but also let us look at logs from more than just the past few days, at the same time. If a book should only be rebuilt about monthly, 100 days would let us look back 2-3 months or so for any given book.

**Comment by eshellman** [comment]
I cycle the rebuild.log and ebookconverter.log files monthly. (I run conversion_summary.sh before zipping the month's ebookconverter.log.
This could be automated; but I've been studying the files created by conversion_summary offline.
The per book logs written by ebookmaker go back to June of 2015 and were previously quite verbose; 1 year of backfiles is useful and sufficient.
a reaper is overkill and possible harmful.

**Comment by eshellman** [comment]
an example: https://www.gutenberg.org/cache/epub/3564/pg3564.converter.log

**Comment by tangledhelix** [comment]
> a reaper is overkill and possible harmful

I'm not sure what you're saying there - though at face value I'll take it as saying we should not make a reaper to reap older log files.

I view the current situation as a problem in two ways, which is why I suggested per-run log files that would thus be easy to clean up later.

1. Since the current per-book logs are single files, removing old contents is a messy affair. We would need either a parsing program or a "smart grep" at minimum, in order to discard old logs without losing new logs, due to their colocation within one file.
2. The current arrangement also adds unnecessary i/o load on the conversion process, and this i/o load accumulates like compounding interest month-over-month. In order to append to a long file, the system has to seek to the end of the file after opening the filehandle. This adds time to the process without benefit.

**Comment by cpeel** [comment]
I agree with Dan that the current situation is problmatic but perhaps from a different angle.

Systems in production need logging, but logging that is usable and maintainable. Unbounded logs, in size or quantity, aren't usable.

Any system that requires a human to go in and do something manually to maintain log files is not a good practice.
